pyogrio/raw.py

```diff
@@ -527,6 +524,67 @@ def _preprocess_options_kwargs(driver, dataset_options, layer_options, kwargs):
     return dataset_kwargs, layer_kwargs
 
 
+def _get_write_path_driver(path, driver, append=False):
+    """Validate and return path and driver
+
+    Parameters
+    ----------
+    path : str or io.BytesIO
+        path to output file on writeable file system or an io.BytesIO object to
+        allow writing to memory
+    driver : str, optional (default: None)
+        The OGR format driver used to write the vector file. By default attempts
+        to infer driver from path.  Must be provided to write to a file-like
+        object.
+    append : bool, optional (default: False)
+        True if path and driver is being tested for append support
+
+    Returns
+    -------
+    (path, driver)
+    """
+
+    if isinstance(path, BytesIO):
+        if driver is None:
+            raise ValueError("driver must be provided to write to in-memory file")
+
+        # blacklist certain drivers known not to work in current memory implementation
+        # because they create multiple files
+        if driver in {"ESRI Shapefile", "OpenFileGDB"}:
+            raise ValueError(f"writing to in-memory file is not supported for {driver}")
+
+        # verify that driver supports VSI methods
+        if not ogr_driver_supports_vsi(driver):
+            raise DataSourceError(
+                f"{driver} does not support ability to write in-memory in GDAL "
+                f"{get_gdal_version_string()}"
+            )
+
+        if append:
+            raise NotImplementedError("append is not supported for in-memory files")
+
+    else:
+        path = str(path)
```

Comment:
Should this still call `vsi_path(..)`? (what it did before in the deleted code below)

Reply:
I took it out because I wasn't convinced that it would work properly and my instead emit confusing errors (though stringifying something that would have gotten a VSI path would be confusing too).  We don't have any tests that prove writing to `/vsi*/` paths would be successful.  I also didn't see support for writing to a `/vsi*/` path in Fiona; they are largely used for reading.

For instance, it looks like it is not possible to write via the `/vsizip/` handler:

```Python
write_dataframe(df, "/vsizip//tmp/test.shp.zip", driver="ESRI Shapefile")

Traceback (most recent call last):
...
pyogrio.errors.DataSourceError: Driver ESRI Shapefile does not support write functionality
```

Are you aware of any things that we can write to a `/vsi*/` path that should succeed?  I can also restore the original behavior, in case our current unknown / untested functionality is working for someone.

Reply:
In theory `/vsizip` should have write capabilities (https://gdal.org/user/virtual_file_systems.html#write-capabilities), but indeed I also don't get that working (maybe the way we handle that path internally isn't fit for working with such a vsi path)

But something like writing to S3 with `vsis3` should in theory also be supported? (although that is indeed something we don't test)

Reply:
I quickly tested this with a public test bucket, and I can write to an "s3://" path (which will get converted to /vsis3 by our preprocessing)

Reply:
Thanks for testing!  That confirms we should indeed leave it in place for now.